receiver/prometheusoperatorreceiver/DESIGN.md

```diff
@@ -0,0 +1,164 @@
+    namespaces:
+      names:
+      - ingress
+```
+
+Sharding should be supported using the optional `sharding` options, which declares which to which shard the receiver
+belongs to. Details about sharding can be found in the [sharding](#sharding) section. 
+After generation of the base config the shard substitution `$(SHARD)` is replaced  with the current shard instance 
+(default = `1`)
+
+This config is then used to create a `prometheusreceiver` `Config` struct.  
+
+### Major components of the config generation
+- **[ConfigGenerator](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L304):**
+  PrometheusOperator component which generates a valid Prometheus config marshalled to a byte array
+- **[ConfigLoader](https://github.com/prometheus/prometheus/blob/main/config/config.go#L68):**
+  Prometheus configuration loader which unmarshalls the config to a `prometheusreceiver` usable object 
+
+
+## Processing config change events
+The Prometheus config is at first compared against the currently applied one. Should there be any change, a new 
+`prometheusreceiver` is started using the generated configuration. 
+If the startup is successful the old instance of `prometheusreceiver` is shutdown. 
+Should any error occur during startup the old instance will keep running
+
+## Collector vs Agent deployment
+The receiver should support both reference architectures of the collector. 
+
+### Collector
+If running as collector the Prometheus config provided by PrometheusOperator can be reused without a change.    
+Should multiple instances with the same config run in the same cluster, they will act like a 
+high availability pair of Prometheus. Therefore, all targets will be scraped multiple times and telemetry 
+will have to deduplicated/compacted later on. 
+
+```yaml
+receivers:
+  prometheus_operator:
+    namespaces: []
+    monitor_selector:
+      match_labels:
+        prometheus-operator-instance: a-instance
+```
+
+### Agent
+In this case the collector is deployed as agent. The receiver can be limited to workloads on a single node
+using the `limit_to_node` option and adding the node name as an environment variable.
+
+```yaml
+env:
+  - name: K8S_NODE_NAME
+    valueFrom:
+      fieldRef:
+        fieldPath: spec.nodeName
+```
+
+```yaml
+receivers:
+  prometheus_operator:
+    namespaces: []
+    limit_to_node: ${K8S_NODE_NAME}
+    monitor_selector:
+      match_labels:
+        prometheus-operator-instance: a-instance
+```
+
+## Sharding
+Sharding is implemented trough sharding the targets which the receiver is scraping.    
+
+This is achieved through `hashmod` relabeling rules which are provided by the `ConfigGenerator`, which implements 
+sharding based onto the `__address__` label.    
+
+This necessitates additional optional config options:
+```yaml
+sharding:
+  shard_count: 1 # number of total shards 
+  shard: 1   # current shard of the receiver.   1 <= value <= shard_count 
```

Comment:
Does this require the use of a distinct configuration source per instance?  Is there a way to avoid that?

Reply:
I expect the user to use substitutions for this something like this could be stored in a config map and then inject the shard to use per environment variables. This would mimic how the `limit_to_node` option or the `k8observer` is it doing it. 

``` yaml
receivers:
  prometheusoperator:
       sharding:
          shard_count: 2
          shard: ${SHARD}
```